src/hooks/UseForm/useForm/useForm.ts

```diff
@@ -0,0 +1,458 @@
+    () =>
+      !!touched &&
+      touchedAtSomeVertex(fields =>
+        fields ? Object.values(fields).some(fieldHasBeenTouched => fieldHasBeenTouched) : false,
+      ),
+    [touched, touchedAtSomeVertex],
+  )
+  const hasChangedWithoutSubmit = useMemo(
+    () =>
+      !!changed &&
+      changedAtSomeVertex(fields =>
+        fields ? Object.values(fields).some(fieldHasBeenChanged => fieldHasBeenChanged) : false,
+      ),
+    [changed, changedAtSomeVertex],
+  )
+
+  const updateFieldsVertex = useCallback(
+    (
+      fieldName: keyof TData,
+      updater: (
+        current?: Fields<TData, TDimensions>[keyof TData],
+      ) => Fields<TData, TDimensions>[keyof TData],
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) =>
+      fieldsGraph.current.updateVertex(
+        fields =>
+          fields
+            ? { ...fields, [fieldName]: updater(fields?.[fieldName]) }
+            : ({ [fieldName]: updater() } as Fields<TData, TDimensions>),
+        coordinates,
+      ),
+    [],
+  )
+
+  const setFieldsVertex = useCallback(
+    <TFieldName extends keyof TData>(
+      fieldName: TFieldName,
+      newValue: Fields<TData, TDimensions>[TFieldName],
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) =>
+      fieldsGraph.current.updateVertex(
+        fields =>
+          fields
+            ? { ...fields, [fieldName]: newValue }
+            : ({ [fieldName]: newValue } as Fields<TData, TDimensions>),
+        coordinates,
+      ),
+    [],
+  )
+
+  const touch = useCallback(
+    (
+      fieldName: keyof TData,
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) => {
+      updateFieldsVertex(fieldName, fields => ({ ...fields!, hasBeenTouched: true }), coordinates)
+      updateTouchedVertex(fields => ({ ...fields!, [fieldName]: true }), coordinates)
+    },
+    [updateTouchedVertex, updateFieldsVertex],
+  )
+
+  const change = useCallback(
+    (
+      fieldName: keyof TData,
+      coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>,
+    ) => {
+      updateFieldsVertex(fieldName, fields => ({ ...fields!, hasBeenChanged: true }), coordinates)
+      updateChangedVertex(fields => ({ ...fields!, [fieldName]: true }), coordinates)
+    },
+    [updateChangedVertex, updateFieldsVertex],
+  )
+
+  const resetChanged = useCallback(() => {
+    fieldsGraph.current.updateAllVertices(fields =>
+      fields
+        ? Object.keys(fields).reduce(
+            (acc, fieldName) => ({
+              ...acc,
+              [fieldName]: { ...fields[fieldName], changed: false },
+            }),
+            {},
+          )
+        : null,
+    )
+    updateAllChangedVertices(fields =>
+      fields
+        ? Object.keys(fields).reduce((acc, fieldName) => ({ ...acc, [fieldName]: false }), {})
+        : null,
+    )
+  }, [updateAllChangedVertices])
```

Comment:
Errors, touched, and changed, are all tracked in our internal state that does not cause rerenders so that we can access them freely in the functions returned by `register` without rerenders. The values are then updated in the user-exposed graphs only when we want to trigger a rerender.